### PR TITLE
edgeql: Add more library functions.

### PR DESCRIPTION
**Pull request by vpetrovykh**
Add more generic functions to `std` module like min, max, round, etc.

Add some specialized functions to `math` module.

### CODE REVIEW

File: edb/lib/std/30-strfuncs.eql [hunk 1]
```diff
@@ -19,11 +19,84 @@
 ## String functions
 
 
-# std::lower
-# ----------
+CREATE FUNCTION
+std::ascii(val: std::str) -> std::int64
```
Comment (elprans) [inline edb/lib/std/30-strfuncs.eql]: Given that this returns an integer Unicode codepoint, this should be called `ord()`.

File: edb/lib/std/30-strfuncs.eql [hunk 2]
```diff
@@ -19,11 +19,84 @@
...
+
+CREATE FUNCTION
+std::upper(s: std::str) -> std::str
+{
+    FROM SQL FUNCTION 'upper';
+};
+
+
+CREATE FUNCTION
+std::title(s: std::str) -> std::str
```
Comment (elprans) [inline edb/lib/std/30-strfuncs.eql]: I think spelling this out as `titlecase()` would be clearer. Likewise for `lower()` and `upper()`.

Comment (1st1) [reply edb/lib/std/30-strfuncs.eql]: I think they all should have "str_" prefix. I'm -0 on the "case" suffix.

File: edb/lib/std/30-strfuncs.eql (hunk 1 above)
Comment (vpetrovykh) [reply edb/lib/std/30-strfuncs.eql]: Done

File: edb/lib/std/30-strfuncs.eql (hunk 2 above)
Comment (vpetrovykh) [reply edb/lib/std/30-strfuncs.eql]: I've added the "str_" prefix.

File: edb/lib/std/30-strfuncs.eql [hunk 3]
```diff
@@ -19,11 +19,84 @@
...
+{
+    FROM SQL FUNCTION 'initcap';
+};
+
+
+CREATE FUNCTION
+std::lpad(s: std::str, n: std::int64, fill: std::str=' ') -> std::str
```
[the remark this answers is not in the document]
Comment (vpetrovykh) [reply edb/lib/std/30-strfuncs.eql]: Done.

File: edb/lib/math.eql [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+math::variance(vals: SET OF std::anyfloat) -> OPTIONAL std::float64
+{
+    FROM SQL FUNCTION 'variance';
+};
+
+
+# math::var_pop
+# -------------
+CREATE FUNCTION
+math::var_pop(vals: SET OF std::decimal) -> OPTIONAL std::decimal
```
[the remark this answers is not in the document]
Comment (vpetrovykh) [reply edb/lib/math.eql]: I've renamed the "variance" to "var" as that is the typical way it's written in statistics formulas.

**Comment by vpetrovykh** [comment]
I've annotated the x-fails with more details comments about the observed failure. Please take a look at them (you can search for "BUG").
@elprans, the last test in the file isolates the bug I was showing to you in the office.

File: edb/lib/std/30-strfuncs.eql [hunk 5]
```diff
@@ -91,3 +91,86 @@ std::`>` (l: std::str, r: std::str) -> std::bool
 CREATE INFIX OPERATOR
 std::`>=` (l: std::str, r: std::str) -> std::bool
     FROM SQL OPERATOR r'>=';
+
+
+CREATE FUNCTION
+std::ord(val: std::str) -> std::int64
```
Comment (1st1) [inline edb/lib/std/30-strfuncs.eql]: I'd drop `ord()` and `chr()`.  I only used them a few times in my life and I fail to come up with a good example of how they can be used in EdgeQL.

File: edb/lib/std/30-strfuncs.eql [hunk 6]
```diff
@@ -91,3 +91,86 @@ std::`>` (l: std::str, r: std::str) -> std::bool
...
+
+CREATE FUNCTION
+std::str_trim(s: std::str, tr: std::str=' ') -> std::str
+{
+    FROM SQL FUNCTION 'btrim';
+};
+
+
+CREATE FUNCTION
+std::str_reverse(s: std::str) -> std::str
```
Comment (1st1) [inline edb/lib/std/30-strfuncs.eql]: Let's have a polymorphic `reverse()` function that works on bytes, strings, and arrays.

File: edb/lib/std/20-genericfuncs.eql [hunk 7]
```diff
@@ -103,3 +103,108 @@ std::random() -> std::float64
...
+    SET initial_value := False;
+    FROM SQL FUNCTION 'bool_or';
+};
+
+
+# std::round
+# ----------
+
+CREATE FUNCTION
+std::round(val: std::decimal) -> std::decimal
```
[the remark this answers is not in the document]
Comment (1st1) [reply edb/lib/std/20-genericfuncs.eql]: @vpetrovykh this still needs to be addressed.

File: edb/lib/std/30-strfuncs.eql (hunk 6 above)
Comment (vpetrovykh) [reply edb/lib/std/30-strfuncs.eql]: There's no function for reversing bytea or arrays in Postgres. How much sense does the function for reversing a string make? It seems to be a good generalization for languages that deal with iterables and sequences generically, but since we don't have these broad concents in EdgeQL, does it really make sense to reverse strings? There's also the trouble with reversing arbitrary unicode.

File: edb/lib/std/30-strfuncs.eql (hunk 5 above)
Comment (vpetrovykh) [reply edb/lib/std/30-strfuncs.eql]: Dropped.

File: edb/lib/std/30-strfuncs.eql (hunk 6 above)
Comment (1st1) [reply edb/lib/std/30-strfuncs.eql]: Yeah, drop it.

**Comment by 1st1** [comment]
`std::round()` should be `FROM SQL FUNCTION 'round'`; please fix

File: tests/test_edgeql_functions.py [hunk 8]
```diff
@@ -1239,3 +1243,1141 @@ class TestEdgeQLFunctions(tb.QueryTestCase):
...
+            {True},
+            {True},
+            {True},
+            {True},
+        ])
+
+    # BUG: we don't yet validate that the return cardinality is 1 here
+    #
+    # Population standard deviation is not defined for an empty set.
+    @unittest.expectedFailure
```
Comment (1st1) [inline tests/test_edgeql_functions.py]: Please rebase onto the most recent `master` branch and start using `@edb.tools.test.not_implemented(reason)` and `@edb.tools.test.xfail(reason)` decorators.

Comment (vpetrovykh) [reply tests/test_edgeql_functions.py]: I've rebased and updated the tests.

File: edb/lib/std/30-arrayfuncs.eql [hunk 9]
```diff
@@ -36,13 +36,13 @@ std::array_unpack(array: array<anytype>) -> SET OF anytype
 
 
 CREATE FUNCTION
-std::array_contains(array: array<anytype>, what: anytype) -> std::bool
+std::array_contains(haystack: array<anytype>, needle: anytype) -> std::bool
```
[the remark this answers is not in the document]
Comment (elprans) [reply edb/lib/std/30-arrayfuncs.eql]: `array` and `element` seem better.

File: edb/lib/std/20-genericfuncs.eql (hunk 7 above)
[the remark this answers is not in the document]
Comment (vpetrovykh) [reply edb/lib/std/20-genericfuncs.eql]:
```
cannot create the `std::round(val: std::decimal, d: std::int64)` function: overloading "FROM SQL FUNCTION" functions is allowed only when all functions point to the same SQL function
```
Basically, there's a version of `round` where you get to specify the decimal digits and it has to be explicitly written in SQL because it involves a downcast to `int4`. This seems to clash with `FROM SQL FUNCTION 'round';`.

File: tests/test_edgeql_functions.py [hunk 10]
```diff
@@ -1239,3 +1243,965 @@ class TestEdgeQLFunctions(tb.QueryTestCase):
...
+            SELECT math::ceil(<decimal>-2.5);
+        ''', [
+            {2},
+            {3},
+            {-2},
+            {3},
+            {-2},
+        ])
+
+    # BUG: invoking the polymorphic function ceil() with an int
```
[the remark this answers is not in the document]
Comment (vpetrovykh) [reply tests/test_edgeql_functions.py]: Fixed.

File: edb/lib/std/30-strfuncs.eql (hunk 6 above)
Comment (vpetrovykh) [reply edb/lib/std/30-strfuncs.eql]: Dropped.

File: edb/lib/std/30-arrayfuncs.eql (hunk 9 above)
[the remark this answers is not in the document]
Comment (vpetrovykh) [reply edb/lib/std/30-arrayfuncs.eql]: OK, I'm not terribly attached to the haystack/needle. Array/element will do just fine.

**Comment by vpetrovykh** [comment]
I've left a comment regarding `round` in the first comment thread that mentioned it. 

TL;DR the `FROM SQL FUNCTION` is incompatible with the third `round` definition.

File: edb/lib/std/30-arrayfuncs.eql (hunk 9 above)
[the remark this answers is not in the document]
Comment (vpetrovykh) [reply edb/lib/std/30-arrayfuncs.eql]: Hm. Do you object to the haystack/needle terminology in other contexts? I use that for `find`, but I can change that to something else in principle.

File: edb/lib/std/20-genericfuncs.eql (hunk 7 above)
[the remark this answers is not in the document]
Comment (1st1) [reply edb/lib/std/20-genericfuncs.eql]: got it
